File: app/src/main/java/im/fdx/v2ex/network/cookie/SharedPrefsPersistor.java
package im.fdx.v2ex.network.cookie;

import android.content.Context;
import android.content.SharedPreferences;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.Set;

import im.fdx.v2ex.network.NetManager;
import okhttp3.Cookie;

/**
 * Created by fdx on 2017/3/16.
 *
 * 保存cookie 到 preference
 */

public class SharedPrefsPersistor implements CookiePersistor {

    private SharedPreferences sharedPreferences;


    public SharedPrefsPersistor(Context context) {
        this.sharedPreferences = context.getSharedPreferences("v2ex_cookie", Context.MODE_PRIVATE);
    }

    public SharedPrefsPersistor(SharedPreferences sharedPreferences) {
        this.sharedPreferences = sharedPreferences;
    }

    @Override
    public void removeAll(List<Cookie> cookies) {
        SharedPreferences.Editor editor = sharedPreferences.edit();
        for (Cookie cookie : cookies) {
            editor.remove(createCookieKey(cookie));
        }
        editor.apply();

    }

    @Override
    public void clear() {
        sharedPreferences.edit().clear().apply();
    }

    @Override
    public void persist(Cookie cookie) {


    }

    @Override
    public void persistAll(Collection<Cookie> cookies) {

        SharedPreferences.Editor editor = sharedPreferences.edit();
        for (Cookie cookie : cookies) {
//            String encode = new CookiesSerializable().encode(cookie);
            String encode = NetManager.myGson.toJson(cookie);
//            XLog.d("fdx:cookiePersist: " + encode);
            editor.putString(createCookieKey(cookie), encode);
        }
        editor.apply();

    }


    private static String createCookieKey(Cookie cookie) {
        return (cookie.secure() ? "https" : "http") + "://" + cookie.domain() + cookie.path() + "|" + cookie.name();
    }

    @Override
    public List<Cookie> loadAll() {

        List<Cookie> cookies = new ArrayList<>();

        Set<? extends Map.Entry<String, ?>> mapSet = sharedPreferences.getAll().entrySet();
        for (Map.Entry<String, ?> stringEntry : mapSet) {
            String strCookie = (String) stringEntry.getValue();
//            Cookie cookie = new CookiesSerializable().decode(strCookie);

            Cookie cookie = NetManager.myGson.fromJson(strCookie, Cookie.class);
            cookies.add(cookie);
        }

        return cookies;
    }

    private Cookie getDecode(String strCookie) {
        return new CookiesSerializable().decode(strCookie);
    }
}

File: app/src/main/java/im/fdx/v2ex/network/cookie/SharedPrefsPersistor.kt
package im.fdx.v2ex.network.cookie

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import com.google.gson.Gson
import im.fdx.v2ex.network.NetManager
import im.fdx.v2ex.utils.extensions.loge
import im.fdx.v2ex.utils.extensions.logv
import okhttp3.Cookie
import java.util.*

/**
 * Created by fdx on 2017/3/16.

 * 保存cookie 到 preference
 */

class SharedPrefsPersistor : CookiePersistor {

    private var sharedPreferences: SharedPreferences

    constructor(context: Context) {
        this.sharedPreferences = context.getSharedPreferences("v2ex_cookie", Context.MODE_PRIVATE)
    }

    constructor(sharedPreferences: SharedPreferences) {
        this.sharedPreferences = sharedPreferences
    }

    override fun removeAll(cookies: List<Cookie>) {
        val editor = sharedPreferences.edit()
        for (cookie in cookies) {
            editor.remove(createCookieKey(cookie))
        }
        editor.apply()

    }

    override fun clear() {
        sharedPreferences.edit().clear().apply()
    }

    override fun persist(cookie: Cookie) {

    }

    override fun persistAll(cookies: Collection<Cookie>) {
        val editor = sharedPreferences.edit()
        for (cookie in cookies) {
            val encode =  Gson().toJson(cookie)
            editor.putString(createCookieKey(cookie), encode)
        }
        editor.apply()

    }


    private fun createCookieKey(cookie: Cookie): String {
        return "${if (cookie.secure) "https" else "http"}://${cookie.domain}${cookie.path}|${cookie.name}"
    }


    @Deprecated("目前没有用到")
    override fun loadAll(): List<Cookie> {

        val cookies = ArrayList<Cookie>()

        val mapSet = sharedPreferences.all.entries
        for ((_, value) in mapSet) {
            val strCookie = value as String
            val cookie = NetManager.myGson.fromJson(strCookie, Cookie::class.java)
            cookies.add(cookie)
        }
        return cookies
    }


    fun loadByHost(url: String): List<Cookie> {
        val cookies = ArrayList<Cookie>()
        val mapSet = sharedPreferences.all.entries
        mapSet.forEach { (key, value) ->
            if (key.contains(url)) {
                val strCookie = value as String
//                logv("fdx2:$strCookie")
                val cookie = NetManager.myGson.fromJson(strCookie, Cookie::class.java)
                cookies.add(cookie)
            }
        }
        return cookies
    }
}
